Add support for GD32F307VG

ChazJin opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
This pull request adds support for GD32F307VG MCU and GD EMAC driver. All the tests were verified on GD32-F307VG board ("detect_code": ["1701"]).

#8713  is the old version and I have closed it. In contrast to the older version, we added the licence of the files, modified some file names, and added the GD EMAC driver files.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

ChazJin committed:
Add GD32F307VG

ChazJin committed:
Add GD32F307VG

ChazJin committed:
Re add target support for GD32F307VG

1. Mainly change TARGET_Gigadevice --> TARGET_GigaDevice
2. Add license header

ChazJin committed:
Re add target support for GD32F307VG

1. Mainly change TARGET_Gigadevice --> TARGET_GigaDevice
2. Add license header

ChazJin committed:
Delete added information (GD32F307VG)

ChazJin committed:
Delete added information (GD32F307VG)

ChazJin committed:
Add EMAC driver for GD32_F307VG

ChazJin committed:
Add EMAC driver for GD32_F307VG

ChazJin committed:
Update licence

ChazJin committed:
Update licence

ChazJin committed:
Update licence

ChazJin committed:
Update licence

ChazJin committed:
Update licence

ChazJin committed:
Update licence

ashok-rao commented on targets/TARGET_GigaDevice/TARGET_GD32F30X/i2c_api.c:
```diff
@@ -0,0 +1,653 @@
+}
+
+/** Reset I2C peripheral. TODO: The action here. Most of the implementation sends stop()
+ *
```
Anything pending here?

ChazJin replied:
The comment is copied from \mbed-os\hal\i2c_api.h line 126 and we refer to it.

ashok-rao commented on targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/PinNames.h:
```diff
@@ -0,0 +1,239 @@
+    SPI_MISO    = PORTA_6,
+    SPI_SCK     = PORTA_5,
+    SPI_CS      = PORTE_3,
+    PWM_OUT     = PORTA_7,
```
Only 1 of each peripheral brought out on the board?

ChazJin replied:
Yes, there are one SPI-Flash, one EEPROM on the board. The MCU has more than one peripheral, and meanwhile the pins are all out on the board.

ChazJin commented:
The test results are as follows(WIN10, run "mbed test -t GCC_ARM/ARM/IAR -m GD32_F307VG" directly):
[GD32_F307VG_ARM.txt](https://github.com/ARMmbed/mbed-os/files/2580600/GD32_F307VG_ARM.txt)
[GD32_F307VG_GCC.txt](https://github.com/ARMmbed/mbed-os/files/2580601/GD32_F307VG_GCC.txt)
[GD32_F307VG_IAR.txt](https://github.com/ARMmbed/mbed-os/files/2580602/GD32_F307VG_IAR.txt)

adbridge commented:
@ChazJin Could you please rebase this ?

ChazJin committed:
Add support for GD32F307VG

Add GD32F307VG MCU information to target.json

ChazJin committed:
Add support for GD32F307VG

Add GD32F307VG MCU information to target.json

ChazJin committed:
Delete a button definition

ChazJin committed:
Delete a button definition

ChazJin commented:
> @ChazJin Could you please rebase this ?

Hello @adbridge, I have rebased it and keep my own repository(https://github.com/ChazJin/mbed-os) consistent with the remote repository(https://github.com/ARMmbed/mbed-os). Would you please let me know is it the right way of doing it?

My previous tests were based on 5.10.1, and I will upload the test results based on the latest 5.10.3 library later.

adbridge commented:
@ChazJin  we seem to be getting a number of failures:

```
Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_GCC_ARM/startup_gd32f30x_cl.s
Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_ARM_MICRO/startup_gd32f30x_cl.s
Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_ARM_STD/startup_gd32f30x_cl.s
Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_IAR/startup_gd32f30x_cl.s
```
```
           AssertionError: Target GD32_F307VG contains invalid device_name GD32F307VG
E           assert 'GD32F307VG' in {'A31G111': {'algorithm': {'A31G11x/Flashloader/A31G11x_series_CFG.FLM': {'default': '1', 'ramsize': None, 'ramstart':...': 'A31G12x/Core/include/A31G12x_series.h'}, 'core': 'Cortex-M0', 'debug': 'A31G12x/SVD/A31G12x_series.svd', ...}, ...}
E            +  where 'GD32F307VG' = Target(name=u'GD32_F307VG', json_data={u'GD32_Target': OrderedDict([(u'inherit...rget', 2)], resolution_order_names=[u'GD32_F307VG', u'GD32_Target', u'Target']).device_name
E            +  and   {'A31G111': {'algorithm': {'A31G11x/Flashloader/A31G11x_series_CFG.FLM': {'default': '1', 'ramsize': None, 'ramstart':...': 'A31G12x/Core/include/A31G12x_series.h'}, 'core': 'Cortex-M0', 'debug': 'A31G12x/SVD/A31G12x_series.svd', ...}, ...} = <tools.arm_pack_manager.Cache instance at 0x7f88c9a3f6c8>.index
```
@ashok-rao can you help?

ashok-rao commented:
Where is this "GD32_F307VG" (with the underscore) coming from? from the logs above, I see: "where 'GD32F307VG' = Target(name=u'GD32_F307VG'," 

Also, there seems to be a typo TARGET_GD32F30X/TA **T** GET_GD32F307VG .. not entirely sure if fixing these would help?

ChazJin commented:
> @ChazJin we seem to be getting a number of failures:
> 
> ```
> Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_GCC_ARM/startup_gd32f30x_cl.s
> Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_ARM_MICRO/startup_gd32f30x_cl.s
> Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_ARM_STD/startup_gd32f30x_cl.s
> Bad Assembler file name found: ./targets/TARGET_GigaDevice/TARGET_GD32F30X/TATGET_GD32F307VG/device/TOOLCHAIN_IAR/startup_gd32f30x_cl.s
> ```
> ```
>            AssertionError: Target GD32_F307VG contains invalid device_name GD32F307VG
> E           assert 'GD32F307VG' in {'A31G111': {'algorithm': {'A31G11x/Flashloader/A31G11x_series_CFG.FLM': {'default': '1', 'ramsize': None, 'ramstart':...': 'A31G12x/Core/include/A31G12x_series.h'}, 'core': 'Cortex-M0', 'debug': 'A31G12x/SVD/A31G12x_series.svd', ...}, ...}
> E            +  where 'GD32F307VG' = Target(name=u'GD32_F307VG', json_data={u'GD32_Target': OrderedDict([(u'inherit...rget', 2)], resolution_order_names=[u'GD32_F307VG', u'GD32_Target', u'Target']).device_name
> E            +  and   {'A31G111': {'algorithm': {'A31G11x/Flashloader/A31G11x_series_CFG.FLM': {'default': '1', 'ramsize': None, 'ramstart':...': 'A31G12x/Core/include/A31G12x_series.h'}, 'core': 'Cortex-M0', 'debug': 'A31G12x/SVD/A31G12x_series.svd', ...}, ...} = <tools.arm_pack_manager.Cache instance at 0x7f88c9a3f6c8>.index
> ```
> @ashok-rao can you help?

@adbridge Is this related to our current MCU being temporarily not in CMSIS PACK? And our engineers have modified the keil5 pack of GigaDevice MCU, which will be uploaded to the server and will take effect soon. FLM and SVD files of GigaDevice MCU will definitely be found next week at the latest.

theotherjimmy commented:
@ChazJin The first error is that your assembly files end with `.s`. We require (for maximum compatibility) that they end with `.S` (it's a capitol instead of a small).

The second error is that your MCU is not in CMSIS Pack yet. You can remove `device_name` from your target in `targets.json` to get this in faster, and later submit a PR that adds back that value with the data from the released CMSIS Pack.

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Add it to replace startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Remove the "device_name": "GD32F307VG"

Our MCU is not in CMSIS Pack and later I will submit a PR that adds back the value with the data from the released CMSIS Pack.

ChazJin commented:
> @ChazJin The first error is that your assembly files end with `.s`. We require (for maximum compatibility) that they end with `.S` (it's a capitol instead of a small).
> 
> The second error is that your MCU is not in CMSIS Pack yet. You can remove `device_name` from your target in `targets.json` to get this in faster, and later submit a PR that adds back that value with the data from the released CMSIS Pack.

@theotherjimmy Thank you very much for your solution. And I have uploaded the new file and removed the `device name`.

@adbridge Please help us to see if there are any other failures after this modification when you are avilable. Thanks a lot.

adbridge commented:
@ChazJin Unfortunately you have made a merge commit to get your changes in. Merge commits cause problems when trying to backport to other branches for patch releases. Could you please use the rebase command instead. Thanks.

ChazJin committed:
Consistent with the remote repo and add GD32F307VG

ChazJin committed:
conflicts resolve

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Delete startup_gd32f30x_cl.s

ChazJin committed:
Delete added information (GD32F307VG)

ChazJin committed:
Error modification

0xc0170 commented on targets/TARGET_GigaDevice/TARGET_GD32F30X/mbed_overrides.c:
```diff
@@ -0,0 +1,154 @@
+
+    SystemCoreClockUpdate();
+
+    mbed_sdk_inited = 1;
```
This access does not need to be thread/interupt safe (this is used in the HAL as I've seen)

ChazJin replied:
‘mbed_sdk_inited’ is used in our firmware library for obtaining hardware delay at startup. It's no problem to delete it, but it is not conducive to our later expansion. 

So the question is, SHOULD it be deleted from our firmware library? Thanks a lot. @0xc0170

0xc0170 replied:
It can stay, I was just curious is access like this is fine (no race conditions that might lead do undefined behavior)

ChazJin replied:
"mbed_sdk_inited" is used at different stages of MBED initialization to calculate hardware latency based on different system clocks, for example, at the beginning of boot ("mbed_sdk_inited" equals to 0) the system clock is 8 MHz and at the end of boot ("mbed_sdk_inited" equals to 1) the system clock may be 72 MHz. In this case, if 10ms delay is required, two methods of calculating delay are needed because the system clock is different. 

So we just use this variable as a symbol and we think do not delete it for the time being. We should like to ask is it a question resolved?

0xc0170 replied:
All good.

0xc0170 commented on targets/TARGET_GigaDevice/mbed_rtx.h:
```diff
@@ -0,0 +1,41 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
```
Can you add SPDX identifier to new files (like HAL implementation , this config file).

```
/*
 * Copyright (c) 2009-2018 Arm Limited. All rights reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the License); you may
```

ChazJin replied:
Our licence statement is based on the standard format of Apache-2.0 and BSD-3-Clause.  If they are modified, it would be different from the standard format. Therefore, our puzzle is whether it's okay if it's inconsistent with the standard statement, and how to modify it. Are there any standards or documents to be revised for our reference?

We considered how to revise it ourselves and there are a few questions that need to be confirmed.

1. Files in \targets\TARGET_GigaDevice\TARGET_GD32F30X\GD32F30x_standard_peripheral are our standard peripheral library. The license of these files will be changed to the following, could you please help us confirm this change?
```
/*
    Copyright (c) 2018, GigaDevice Semiconductor Inc. '*All rights reserved.*'

    '*SPDX-License-Identifier: BSD-3-Clause*'

    Redistribution and use in source and binary forms, with or without modification, 
```
2. The license of other files that follow Apache-2.0 licence will be changed to the following, and it is needed to be confirmed too.
```
/* 
 * Copyright (c) '*2009-*'2018 GigaDevice Semiconductor Inc. '*All rights reserved.*'
 *
 * '*SPDX-License-Identifier: Apache-2.0*'
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
```
**NOTE**: The text between '* *' or italics as we can see is the modified content compared to current file.

0xc0170 replied:
I was requesting files that contain Mbed OS porting code (like HAL us ticker, rtc and similar), not changing your SDK (peripheral library). We should use SPDX identifiers where we can. 

1st: dont need to change your peripheral library
2nd: files that contain Mbed OS porting should contain SPDX identifier

Does this clear it or ?

ChazJin replied:
Thank you for your detailed reply. Maybe we have referenced to other files and found other manufacturers' files haven't changed. Therefore, it is not clear how to add the SPDX identifier for us. And we will add a new line for "SPDX-License-Identifier: Apache-2.0" in the head of the licence declaration.

0xc0170 reviewed (changes requested):
One request change and a question

ChazJin committed:
Add SPDX identifier

ChazJin committed:
Add SPDX identifier for GD EMAC driver

0xc0170 commented:
@ashok-rao Can you finalize the review please?

ashok-rao reviewed (approved):
LGTM.. Happy to proceed with this. Thanks @ChazJin

cmonr commented:
CI started.

0xc0170 commented:
This is the build failure

`[Error] mbed_mpu_v7m.c@26,0:  #35: #error directive: "Device has v7m MPU but it is not enabled. Add 'MPU' to device_has in targets.json"`

You should be able to reproduce this locally.

ChazJin committed:
Add MPU label and fix labels alignment

ChazJin commented:
> This is the build failure
> 
> `[Error] mbed_mpu_v7m.c@26,0: #35: #error directive: "Device has v7m MPU but it is not enabled. Add 'MPU' to device_has in targets.json"`
> 
> You should be able to reproduce this locally.

I have added it and please check if there are any other questions. Thanks a lot. @0xc0170

0xc0170 commented:
`13:10:18 [GD32_F307VG:ARM]         Error: L6218E: Undefined symbol flash_get_erase_value (referred from BUILD/tests/GD32_F307VG/ARM/drivers/FlashIAP.o).
` latest failure, please test locally ARM/GCC ARM/IAR

ChazJin committed:
Error resolve, bug fix and 2 new targets information add

1. Resolve the Error: L6218E: Undefined symbol flash_get_erase_value
2. Bug fix (ADC, USART, etc.)
3. Modify the wrong name (TATGET_GigaDevice-->TARGET_GigaDevice)
4. Add new targets (GD32-F450ZI and GD32-E103VB) support to target.json
5. Add INITIAL_SP macro for GD32-F450ZI and GD32-E103VB

ChazJin commented:
> `13:10:18 [GD32_F307VG:ARM] Error: L6218E: Undefined symbol flash_get_erase_value (referred from BUILD/tests/GD32_F307VG/ARM/drivers/FlashIAP.o). ` latest failure, please test locally ARM/GCC ARM/IAR

Hi @0xc0170 , we have added the impletation of flash_get_erase_value and I think Error: L6218E must be resolved. After that, the test results based on MBED OS 5.10.4 are as follows(WIN10, run "mbed test -t GCC_ARM/ARM/IAR -m GD32_F307VG" directly).
[GD32_307VG_ARM_5_10_4.txt](https://github.com/ARMmbed/mbed-os/files/2652958/GD32_307VG_ARM_5_10_4.txt)
[GD32_307VG_GCC_ARM_5_10_4.txt](https://github.com/ARMmbed/mbed-os/files/2652959/GD32_307VG_GCC_ARM_5_10_4.txt)
[GD32_307VG_IAR_5_10_4.txt](https://github.com/ARMmbed/mbed-os/files/2652960/GD32_307VG_IAR_5_10_4.txt)

0xc0170 commented:
CI restarted

ChazJin commented:
> # [Test run: FAILED](https://mbed-os.mbedcloudtesting.com/job/mbed-os-ci/job/PR-8725/11/)
> Summary: 3 of 7 test jobs failed
> Build number : 11
> [Build artifacts](http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/artifacts/8725/)
> 
> Failed test jobs:
> 
> * jenkins-ci/mbed-os-ci_build-GCC_ARM
> * jenkins-ci/mbed-os-ci_build-ARM
> * jenkins-ci/mbed-os-ci_build-IAR

I'm sorry to say that we can't open these links above and the Details of jenkins-ci/build-ARM/GCC_ARM/IAR, so we can't see the specific errors. Could you please help us? @0xc0170 Thank a lot!

cmonr commented:
@ChazJin Do you have access to this page? http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/artifacts/8725/11/

It requires javascript (not sure why...), if that helps.

ChazJin committed:
Delete GD32F450ZI and GD323103VB which are not implemented

Delete GD32F450ZI and GD323103VB which are not implemented yet and add them later.

ChazJin commented:
@cmonr Thank you for your advise and I have found the reason for the errors. The "GD32F450ZI" and "GD32E103VB" are the projects we're working on, but they're not finished yet.  After I removed the information about these two targets from target.json, I think the result will be OK. The links are as follows.
http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/artifacts/8725/11/build-ARM/PASS/GD32_F307VG/
http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/artifacts/8725/11/build-ARM/FAIL/GD32_E103VB/
http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/artifacts/8725/11/build-ARM/FAIL/GD32_F450ZI/

@0xc0170 Please help us to run CI test again, and I really appreciate it.

ChazJin committed:
Style Format for GD32F30x standard peripheral files

cmonr commented:
Prioritizing RC3 PRs. Will start when able.

cmonr commented:
CI started

cmonr commented:
@ChazJin We're working on releasing 5.11.1, and are testing the RC now: https://github.com/ARMmbed/mbed-os/pull/9208

However, this PR gave us quite a bit of trouble. When working on future PRs, _please_ make sure that commits from patch resolutions are not in the commit history. Lines like this https://github.com/ARMmbed/mbed-os/pull/8725/commits/5a15f810a604aba56349c956615909e26c9f3b32#diff-e9bd16efde5b7f92c0dcd845c5cf3619L352 caused us to manually apply this PR to the release.

ChazJin commented:
We are VERY SORRY to have caused you so much trouble. This is our first PR and we will surely pay attention to it in the future PRs. Once again we would like to apologize for it. @cmonr
